File: typescript/src/identifier.ts
/**
 * The extra attributes that identify a subscription alongside its channel
 * name, like the id of the record a channel streams for.
 */
export type Params = Record<string, unknown>;

/**
 * Names one subscription. It is encoded as a JSON object and the server treats
 * that encoding as an opaque key, echoing it back on every frame it sends for
 * the subscription.
 *
 * ```ts
 * { channel: "RoomChannel", params: { id: 42 } }
 * ```
 *
 * A channel with no params needs only the name.
 */
export interface Identifier {
  channel: string;
  params?: Params;
}

/**
 * The JSON identifier string the server knows a subscription by, and the one
 * it echoes back on everything it sends for it.
 *
 * The keys are sorted, which is what Go's `json.Marshal` of a map does and so
 * what the Go client puts on the wire. The server treats the string as opaque,
 * so the order only has to be stable — but sorting means the same identifier
 * keys the same subscription whichever client built it.
 */
export function identifierKey(identifier: Identifier): string {
  const fields = { ...identifier.params, channel: identifier.channel };

  try {
    return JSON.stringify(sortedDeeply(fields));
  } catch (cause) {
    throw new TypeError(`actioncable: encoding identifier for ${identifier.channel}`, { cause });
  }
}

function sortedDeeply(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(sortedDeeply);
  } else if (isPlainObject(value)) {
    const sorted: Params = {};
    for (const key of Object.keys(value).sort()) {
      sorted[key] = sortedDeeply(value[key]);
    }
    return sorted;
  } else {
    return value;
  }
}

function isPlainObject(value: unknown): value is Params {
  if (value === null || typeof value !== "object") {
    return false;
  }

  const prototype: unknown = Object.getPrototypeOf(value);

  return prototype === Object.prototype || prototype === null;
}

File: rust/src/identifier.rs
use std::collections::BTreeMap;
use std::fmt;

use serde_json::Value;

/// Names one subscription: a channel and the params that go with it, like the id of the
/// record the channel streams for.
///
/// The server treats the identifier's JSON encoding as an opaque key and echoes it back on
/// every frame it sends for the subscription, so the encoding has to be stable: the keys
/// come out sorted, `channel` among them.
///
/// ```
/// use actioncable::Identifier;
///
/// let room = Identifier::new("RoomChannel").param("id", 42);
/// assert_eq!(r#"{"channel":"RoomChannel","id":42}"#, room.key());
/// ```
///
/// A channel with no params needs only the name.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Identifier {
    channel: String,
    params: BTreeMap<String, Value>,
}

impl Identifier {
    /// Names a channel, with no params.
    pub fn new(channel: impl Into<String>) -> Identifier {
        Identifier {
            channel: channel.into(),
            params: BTreeMap::new(),
        }
    }

    /// Adds one param, like the id of the record the channel streams for.
    pub fn param(mut self, name: impl Into<String>, value: impl Into<Value>) -> Identifier {
        self.params.insert(name.into(), value.into());
        self
    }

    /// The channel's name.
    pub fn channel(&self) -> &str {
        &self.channel
    }

    /// The params, sorted by name as the key encodes them.
    pub fn params(&self) -> &BTreeMap<String, Value> {
        &self.params
    }

    /// The JSON string the server knows the subscription by. A param named `channel` loses
    /// to the channel name.
    ///
    /// Where the Go client returns an error, this cannot fail: a param is a
    /// [`serde_json::Value`], and a map of them always encodes.
    #[allow(
        clippy::expect_used,
        reason = "serializing a map of Values has no failure mode to report"
    )]
    pub fn key(&self) -> String {
        let mut fields = self.params.clone();
        fields.insert("channel".to_string(), Value::String(self.channel.clone()));
        serde_json::to_string(&fields).expect("a map of JSON values always encodes")
    }
}

impl fmt::Display for Identifier {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.key())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_key_is_the_sorted_json_encoding() {
        let identifiers = [
            (
                Identifier::new("RoomChannel"),
                r#"{"channel":"RoomChannel"}"#,
            ),
            (
                Identifier::new("RoomChannel").param("id", 42),
                r#"{"channel":"RoomChannel","id":42}"#,
            ),
            (
                Identifier::new("RoomChannel")
                    .param("since", "yesterday")
                    .param("id", 42),
                r#"{"channel":"RoomChannel","id":42,"since":"yesterday"}"#,
            ),
            (
                Identifier::new("RoomChannel").param("channel", "impostor"),
                r#"{"channel":"RoomChannel"}"#,
            ),
        ];

        for (identifier, key) in identifiers {
            assert_eq!(key, identifier.key());
            assert_eq!(key, identifier.to_string());
        }
    }

    /// The Go client's key can fail on params that don't encode. Here they can't exist: a
    /// param is a `Value`, and a float JSON has no spelling for is one before it arrives.
    #[test]
    fn an_identifier_always_encodes() {
        let identifier = Identifier::new("RoomChannel").param("id", f64::NAN);

        assert_eq!(r#"{"channel":"RoomChannel","id":null}"#, identifier.key());
    }
}
